File: bookmarker-java/src/main/java/com/sivalabs/bookmarker/domain/model/BookmarkDTO.java
package com.sivalabs.bookmarker.domain.model;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.sivalabs.bookmarker.domain.entity.Bookmark;
import com.sivalabs.bookmarker.domain.entity.Tag;
import lombok.Data;

import javax.validation.constraints.NotBlank;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

@Data
public class BookmarkDTO {
    private Long id;

    @NotBlank(message = "URL cannot be blank")
    private String url;

    private String title;

    @JsonProperty("created_user_id")
    private Long createdUserId;

    @JsonProperty("created_user_name")
    private String createdUserName;

    @JsonProperty("created_at")
    private LocalDateTime createdAt;

    @JsonProperty("updated_at")
    private LocalDateTime updatedAt;

    private List<String> tags;

    public List<String> getTags() {
        if(this.tags == null) {
            this.tags = new ArrayList<>();
        }
        return this.tags;
    }

    public static BookmarkDTO fromEntity(Bookmark bookmark) {
        BookmarkDTO dto = new BookmarkDTO();
        dto.setId(bookmark.getId());
        dto.setUrl(bookmark.getUrl());
        dto.setTitle(bookmark.getTitle());
        dto.setCreatedUserId(bookmark.getCreatedBy().getId());
        dto.setCreatedUserName(bookmark.getCreatedBy().getName());
        dto.setCreatedAt(bookmark.getCreatedAt());
        dto.setUpdatedAt(bookmark.getUpdatedAt());
        if(bookmark.getTags() != null) {
            dto.setTags(bookmark.getTags().stream().map(Tag::getName).collect(Collectors.toList()));
        }
        return dto;
    }
}

File: backend/src/main/kotlin/com/sivalabs/bookmarker/domain/model/BookmarkDTO.kt
package com.sivalabs.bookmarker.domain.model

import com.fasterxml.jackson.annotation.JsonProperty
import com.sivalabs.bookmarker.domain.entity.Bookmark
import java.time.LocalDateTime
import javax.validation.constraints.NotBlank

data class BookmarkDTO(
    var id: Long = 0,

    @field:NotBlank(message = "URL cannot be blank")
    var url: String = "",

    var title: String = "",

    @JsonProperty("created_user_id")
    var createdUserId: Long = 0,

    @JsonProperty("created_user_name")
    var createdUserName: String = "",

    @JsonProperty("created_at")
    var createdAt: LocalDateTime = LocalDateTime.now(),

    @JsonProperty("updated_at")
    var updatedAt: LocalDateTime? = LocalDateTime.now(),

    var tags: List<String> = listOf()
) {
    companion object {
        fun fromEntity(bm: Bookmark) = BookmarkDTO(
            bm.id,
            bm.url,
            bm.title,
            bm.createdBy.id,
            bm.createdBy.name,
            bm.createdAt,
            bm.updatedAt,
            bm.tags.map { it.name }
        )
    }
}
